src/wiki/models/article.py

```diff
@@ -189,33 +189,57 @@ class Meta:
             ("grant", _("Can assign permissions to other users")),
         )
 
-    def render(self, preview_content=None):
+    def render(self, preview_content=None, user=None):
         if not self.current_revision:
             return ""
         if preview_content:
             content = preview_content
         else:
             content = self.current_revision.content
         return mark_safe(article_markdown(content, self,
-                                          preview=preview_content is not None))
+                                          preview=preview_content is not None,
+                                          user=user))
 
     def get_cache_key(self):
+        """Returns per-article cache key."""
         lang = translation.get_language()
+
         return "wiki:article:{id:d}:{lang:s}".format(
             id=self.current_revision.id if self.current_revision else self.id,
-            lang=lang
-        )
+            lang=lang)
+
+    def get_cache_content_key(self, user=None):
+        """Returns per-article-user cache key."""
+        return "{key}:{user!s}".format(
+            key=self.get_cache_key(),
+            user=user if user else "")
+
+    def get_cached_content(self, user=None):
+        """Returns cached version of rendered article.
+
+        The cache contains one "per-article" entry plus multiple
+        "per-article-user" entries. The per-article-user entries contain the
+        rendered article, the per-article entry contains list of the
+        per-article-user keys. The rendered article in cache (per-article-user)
+        is used only if the key is in the per-article entry. To delete
+        per-article invalidates all article cache entries."""
 
-    def get_cached_content(self):
-        """Returns cached version of rendered article"""
         cache_key = self.get_cache_key()
-        cached_content = cache.get(cache_key)
-        if cached_content is None:
-            cached_content = self.render()
-            cache.set(cache_key, cached_content, settings.CACHE_TIMEOUT)
-        else:
-            cached_content = mark_safe(cached_content)
-        return cached_content
+        cache_content_key = self.get_cache_content_key(user)
+
+        cached_items = cache.get(cache_key, list())
+
+        if cache_content_key in cached_items:
+            cached_content = cache.get(cache_content_key)
+            if cached_content is not None:
+                return mark_safe(cached_content)
+
+        cached_content = self.render(user=user)
+        cached_items.append(cache_content_key)
+        cache.set(cache_key, cached_items, settings.CACHE_TIMEOUT)
+        cache.set(cache_content_key, cached_content, settings.CACHE_TIMEOUT)
+
+        return mark_safe(cached_content)
```

Comment:
How does this scale? 1000 users = 1000 entries for a front page article?

Reply:
if 1000 users open the page before cache timeout, that yes (all anonymous users = 1 user). For now it is resource wasting, but I am working on plugin that add  `[ifgrp/ifusr ...] group/user specific content [endifgrp/usr]` syntax.

BTW I believe 1000 entries in cache are cheaper that a hash table for find a duplicates, but I don't have numbers.

Reply:
@jenda1 
Thanks for clarifying! :)

I think the cache logic would sit better at the core of the plugin system, so that the plugin designer can mark a specific markdown tag as user-specific. But at least such that cache logic never enters the markdown code itself (user domain!)